src/network/requestQueue/index.js

```diff
@@ -309,15 +316,14 @@ module.exports = class RequestQueue extends EventEmitter {
     // will be fine, and potentially fix up a new timeout if needed at that time.
     // Note that if we're merely "overloaded" by having too many inflight requests
     // we will anyways check the queue when one of them gets fulfilled.
-    let scheduleAt = this.throttledUntil - Date.now()
-    if (!this.throttleCheckTimeoutId) {
-      if (this.pending.length > 0) {
-        scheduleAt = scheduleAt > 0 ? scheduleAt : CHECK_PENDING_REQUESTS_INTERVAL
-      }
+    const timeUntilUnthrottled = this.throttledUntil - this.throttleCurrentTimestamp
```

Comment:
I feel like I'm missing something here. From reading this code it looks to me like we're going to be checking every single time around the event loop, since the `setTimeout` is always called with 0.

So the very first time we get throttled, we'll set `throttleCurrentTimestamp` to 0 and `throttledUntil` to 10. `timeUntilThrottled` is `10 - 0`, which is greater than 0 and we don't already have a scheduled timer, so we schedule it for next tick. Next tick happens so we null out the timeout id and we update `throttleCurrentTimestamp` with the current time, let's say it's 1. Now `timeUntilThrottled` is 10 - 1 which is still greater than 0, so we do the same thing over again.

Reply:
you are right, this would probably create many more calls then necessary, i could imagine that instead using Math.max(timeUntilUnthrottled, 0) would then be more applicable?

Reply:
It might be helpful to read the previous implementation (and the original one). If you check the PR that introduced this (https://github.com/tulios/kafkajs/pull/1532#discussion_r1110809796), I made a similar suggestion. We just need to make sure we don't end up in an infinite loop

Reply:
Is there a way to move this forward? so far it looks good and not causing any infinite loops

Reply:
Ought there to be a test that would guard against this kind of performance degradation to happen in the future?

How we ended up bumping into this was that there was an integration test that did time jumps with Sinon FakeTimer where it was clear that the internals of FakeTimer ended up throttling the actual Event Loop due to this setTimeout (and another setInterval) causing there to be a huge backlog of timer callbacks. Not saying that this is the strategy that one ought to use to guard against this, but that would indirectly create somewhat easy way of verifying that there are no overly-aggressive setTimeout-zero loops in the system.